source/PlayerInfo.cpp

```diff
@@ -2354,7 +2356,35 @@ void PlayerInfo::StepMissions(UI *ui)
 		else if(mission.CanComplete(*this))
 			RemoveMission(Mission::COMPLETE, mission, ui);
 		else if(mission.Destination() == GetPlanet() && !freshlyLoaded)
+		{
 			mission.Do(Mission::VISIT, *this, ui);
+			
+			// On visit dialogs are handled separately as to avoid a player
+			// getting spammed by on visit dialogs if they are stacking jobs
+			// from the same destination.
+			map<Mission::Trigger, MissionAction> actions = mission.GetActions();
+			auto ait = actions.find(Mission::VISIT);
+			if(ait != actions.end())
+			{
+				if(visitText.empty())
+				{
+					map<string, string> subs;
+					subs["<first>"] = firstName;
+					subs["<last>"] = lastName;
+					if(Flagship())
+						subs["<ship>"] = Flagship()->Name();
+					visitText = Format::Replace(ait->second.DialogText(), subs);
```

Comment:
Since `Format::Replace` takes the string by const ref, `DialogText` can return const ref and avoid a string copy.